### PR TITLE
feat: Gateway API support. Fixes #1438

### LINKED ISSUE #1438: Gateway API support for Argo rollouts
[Gateway API](https://gateway-api.sigs.k8s.io/) is a standard out of SIG-Network that evolves the ingress interface with more advanced capabilities such as weight-based load balancing. There are already many [Gateway API implementations](https://gateway-api.sigs.k8s.io/implementations/) that exist including Kong, Traefik, Istio, GKE, and more. This tutorial shows how [traffic splitting](https://gateway-api.sigs.k8s.io/v1alpha1/guides/traffic-splitting/) works in Gateway. 

This issue is to propose Gateway API-based support for Argo rollouts so that any Gateway API implementation can be used with Argo in the same way. 

cc @bowei @robscott

### PR DESCRIPTION
**Pull request by PhilippPlotnikov**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by zachaller** [comment]
I will take a look at reviewing this soon.

**Comment by huikang** [comment]
@PhilippPlotnikov , thanks for the contribution. To help the review the new feature, could you add some description to the doc: https://github.com/argoproj/argo-rollouts/blob/master/docs/features/traffic-management/index.md#how-does-argo-rollouts-enable-traffic-management

File: rollout/trafficrouting/gatewayapi/gatewayapi.go
Comment (huikang) [inline rollout/trafficrouting/gatewayapi/gatewayapi.go:151]: nit: change the parameter name `services` to `rules`

File: rollout/trafficrouting/gatewayapi/gatewayapi.go
Comment (huikang) [inline rollout/trafficrouting/gatewayapi/gatewayapi.go:156]: Could you fix error msg  as it mentions `traefik service`?

File: rollout/trafficrouting/gatewayapi/gatewayapi.go
Comment (huikang) [inline rollout/trafficrouting/gatewayapi/gatewayapi.go:118]: I think we can move the error type of unfound service to the `getService` function

File: rollout/trafficrouting/gatewayapi/gatewayapi.go
Comment (huikang) [inline rollout/trafficrouting/gatewayapi/gatewayapi.go:163]: Shall we `continue` to the next iteration or return?

**Review by huikang — COMMENTED** [review]
Could you add some unit tests to suppress some warnings from the CI? Thanks.

File: rollout/trafficrouting/gatewayapi/gatewayapi.go
Comment (zachaller) [inline rollout/trafficrouting/gatewayapi/gatewayapi.go:110]: Even though this is only checked once  I think along with moving the getService unfound return we could do the same here for backEnd refs not found

File: rollout/trafficrouting/gatewayapi/gatewayapi.go
Comment (zachaller) [reply rollout/trafficrouting/gatewayapi/gatewayapi.go:151]: I think this actually makes more sense as backendRef which maps to https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.HTTPBackendRef if I am reading it correctly it is going through the list of backendRefs and finding the service that it goes to.

File: rollout/trafficrouting/gatewayapi/mocks/gatewayapi.go
Comment (zachaller) [inline rollout/trafficrouting/gatewayapi/mocks/gatewayapi.go:54]: This is not for traefik

File: rollout/trafficrouting/gatewayapi/mocks/gatewayapi.go
Comment (zachaller) [inline rollout/trafficrouting/gatewayapi/mocks/gatewayapi.go:64]: This is not for traefik

File: rollout/trafficrouting/gatewayapi/gatewayapi.go
Comment (zachaller) [inline rollout/trafficrouting/gatewayapi/gatewayapi.go:131]: This is more just a note for myself and for documentation possibly that in gateway api weight is actually 

```
weight/(sum of all weights in this BackendRefs list)

with 0 being send no traffic and a default of 1
```

and that we are assuming a total of 100 which I think is ok

**Comment by shuker85** [comment]
Will this support Zuul Gateway ?

**Comment by shuker85** [comment]
@PhilippPlotnikov hi, i guess it's https://github.com/Netflix/zuul/wiki

**Comment by PhilippPlotnikov** [comment]
@shuker85 am I right it is simple java application ? If yes, you can of course use this feature as Zuul Gateway will be pods simply

**Review by harikrongali — APPROVED** [review]
LGTM

**Comment by harikrongali** [comment]
will wait for @zachaller review before merging it

**Review by todaywasawesome — APPROVED** [review]
LGTM

**Comment by zachaller** [comment]
So I would like to talk about this a bit more based on this https://github.com/argoproj/argo-rollouts/issues/2073. I would like to make this a bit more gitops friendly using the new spec proposed in the issue. As well as a few changes to what is defined in this PR if we find a place to put the service ports argo rollouts would have all the information needed to create and manage the canary route without having to have the user define a route at all. My suggestion is putting the ports on the already existing fields listed below and adding some validation on requiring it for GatewayAPI because it is required for k8s services.

``` 
strategy:
    canary:
      canaryService: argo-rollouts-canary-service:<port>
      stableService: argo-rollouts-stable-service:<port>
```

If the second part of the proposal from the issue is also approved which gives us the ability to add this

```
      trafficRouting:
        managedRoutes:
          - name: primary-route
            canaryRoute: true #This becomes the default/canary route in the upstream traffic router
```

Then we do not even need the config below except for a way to let rollouts know what type of traffic router we are configuring which could be something like `gatewayAPI: {}` or `gatewayAPI: {enabled: true}`  etc
```
trafficRouting:
  gatewayAPI:
    httpRoute: argo-rollouts-http-route # our created httproute
```

**Comment by zachaller** [comment]
So we talked about this a bit more and decided that putting the port into canaryService and stableService would not be a good idea because it breaks customise named references but dropping that config into per traffic router config would be a good approach. 

```
trafficRouting:
  gatewayAPI:
    stableServicePort: 80
    canaryServicePort: 80
```


We also discussed a bit more about the new spec changes and we are liking where they are going but we will probably hold off on adding the part about `canaryRoute: true` into the 1.3 release so that will hold this up for that release. But will get that added to master shortly after.

**Comment by PhilippPlotnikov** [comment]
@zachaller hi. Am I right you would like to add in argo-rollouts feature to create canary-route, stable-route and traffic splitter(in this case gatewayService) with itself so user wont need to create these resources manually, right ?

**Comment by PhilippPlotnikov** [comment]
I didnt understand this
```
      trafficRouting:
        managedRoutes:
          - name: primary-route
            canaryRoute: true #This becomes the default/canary route in the upstream traffic router
```

**Comment by PhilippPlotnikov** [comment]
Im not sure that I understood but it sounds like it needs big change maybe this can be pushed and after I will add another PR with needing update ?

**Comment by mrmm** [comment]
Hello, are there any updates for this PR?

**Comment by zachaller** [comment]
Not much of an update but I will bring this up during community meetings and try to finalize the `canaryRoute`  bit which will then allow this PR to get finished.

**Comment by lucasarrudatrustly** [comment]
Any news about this PR?

**Comment by PhilippPlotnikov** [comment]
Sorry guys, I had some problems, today I will try to resolve conflicts and finish this PR on next week

**Comment by lucasarrudatrustly** [comment]
Any news about this PR?

**Comment by xkkker** [comment]
any updates?

**Comment by rjanovski** [comment]
Flagger [already have this](https://fluxcd.io/flagger/tutorials/gatewayapi-progressive-delivery/)
#just_saying

**Comment by PhilippPlotnikov** [comment]
I am making plugin for GatewayAPI here https://github.com/argoproj-labs/rollouts-gatewayapi-trafficrouter-plugin/pull/1
I hope anything wont be changed and I will soon push it and we will have support for GatewayAPI
Now Im waiting when interface for communication with argo-rollouts will be changed to finish with unit tests and to test work locally

**Comment by zachaller** [comment]
Closing as completed via plugin

**Pull request closed without merging**
